File: tests/bench/poly_dispatch/PolyDispatch.java
// Same program as poly_dispatch.jai.
import java.util.ArrayList;
import java.util.List;

public class PolyDispatch {
    static long scale() {
        String l = System.getenv("BENCH_LEVEL");
        if ("easy".equals(l)) return 16;
        if ("medium".equals(l)) return 4;
        return 1;
    }

    static final long SCALE = scale();
    static final int OPS = 512;
    static final int REPS = (int) Math.max(8000 / SCALE, 1);
    static final long M = 1000003L;

    interface Op {
        long apply(long x);
    }

    static final class AddK implements Op {
        final long k;
        AddK(long k) { this.k = k; }
        public long apply(long x) { return (x + k) % M; }
    }
    static final class MulK implements Op {
        final long k;
        MulK(long k) { this.k = k; }
        public long apply(long x) { return (x * k) % M; }
    }
    static final class SquareK implements Op {
        final long k;
        SquareK(long k) { this.k = k; }
        public long apply(long x) { return (x * x + k) % M; }
    }
    static final class DoubleK implements Op {
        final long k;
        DoubleK(long k) { this.k = k; }
        public long apply(long x) { return (x * 2 + k) % M; }
    }
    static final class DivK implements Op {
        final long k;
        DivK(long k) { this.k = k; }
        public long apply(long x) { return (x + x / k) % M; }
    }
    static final class ModK implements Op {
        final long k;
        ModK(long k) { this.k = k; }
        public long apply(long x) { return (x + x % k) % M; }
    }
    static final class FlipK implements Op {
        final long k;
        FlipK(long k) { this.k = k; }
        public long apply(long x) { return (M - 1 - x + k) % M; }
    }
    static final class MixK implements Op {
        final long k;
        MixK(long k) { this.k = k; }
        public long apply(long x) { return (x + x / 3 + k) % M; }
    }

    static List<Op> build(int n) {
        List<Op> ops = new ArrayList<>();
        long s = 7;
        for (int i = 0; i < n; i++) {
            s = (s * 1103515245L + 12345L) % 2147483648L;
            long which = (s / 65536) % 8;
            long k = s % 97 + 2;
            if (which == 0) ops.add(new AddK(k));
            else if (which == 1) ops.add(new MulK(k));
            else if (which == 2) ops.add(new SquareK(k));
            else if (which == 3) ops.add(new DoubleK(k));
            else if (which == 4) ops.add(new DivK(k));
            else if (which == 5) ops.add(new ModK(k));
            else if (which == 6) ops.add(new FlipK(k));
            else ops.add(new MixK(k));
        }
        return ops;
    }

    public static void main(String[] args) {
        List<Op> ops = build(OPS);
        long acc = 1;
        long check = 0;
        for (int r = 0; r < REPS; r++) {
            acc = (acc + r) % M;
            for (Op op : ops) acc = op.apply(acc);
            check = (check + acc) % M;
        }
        System.out.println(ops.size());
        System.out.println(acc);
        System.out.println(check);
    }
}

File: tests/bench/poly_dispatch/poly_dispatch.cpp
// Same program as poly_dispatch.jai. See tests/bench/README.md for why the C++
// and Java rows are here and what they are not.
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <memory>
#include <vector>

static long bench_scale() {
    const char *l = std::getenv("BENCH_LEVEL");
    if (l && std::strcmp(l, "easy") == 0) return 16;
    if (l && std::strcmp(l, "medium") == 0) return 4;
    return 1;
}
static const long SCALE = bench_scale();
static const int OPS = 512;
static const int64_t M = 1000003;

struct Op {
    int64_t k;
    explicit Op(int64_t k_) : k(k_) {}
    virtual ~Op() {}
    virtual int64_t apply(int64_t x) const = 0;
};

struct AddK : Op {
    explicit AddK(int64_t k) : Op(k) {}
    int64_t apply(int64_t x) const override { return (x + k) % M; }
};
struct MulK : Op {
    explicit MulK(int64_t k) : Op(k) {}
    int64_t apply(int64_t x) const override { return (x * k) % M; }
};
struct SquareK : Op {
    explicit SquareK(int64_t k) : Op(k) {}
    int64_t apply(int64_t x) const override { return (x * x + k) % M; }
};
struct DoubleK : Op {
    explicit DoubleK(int64_t k) : Op(k) {}
    int64_t apply(int64_t x) const override { return (x * 2 + k) % M; }
};
struct DivK : Op {
    explicit DivK(int64_t k) : Op(k) {}
    int64_t apply(int64_t x) const override { return (x + x / k) % M; }
};
struct ModK : Op {
    explicit ModK(int64_t k) : Op(k) {}
    int64_t apply(int64_t x) const override { return (x + x % k) % M; }
};
struct FlipK : Op {
    explicit FlipK(int64_t k) : Op(k) {}
    int64_t apply(int64_t x) const override { return (M - 1 - x + k) % M; }
};
struct MixK : Op {
    explicit MixK(int64_t k) : Op(k) {}
    int64_t apply(int64_t x) const override { return (x + x / 3 + k) % M; }
};

static std::vector<std::unique_ptr<Op>> build(int n) {
    std::vector<std::unique_ptr<Op>> ops;
    int64_t s = 7;
    for (int i = 0; i < n; i++) {
        s = (s * 1103515245LL + 12345LL) % 2147483648LL;
        int64_t which = (s / 65536) % 8;
        int64_t k = s % 97 + 2;
        switch ((int)which) {
            case 0: ops.push_back(std::make_unique<AddK>(k)); break;
            case 1: ops.push_back(std::make_unique<MulK>(k)); break;
            case 2: ops.push_back(std::make_unique<SquareK>(k)); break;
            case 3: ops.push_back(std::make_unique<DoubleK>(k)); break;
            case 4: ops.push_back(std::make_unique<DivK>(k)); break;
            case 5: ops.push_back(std::make_unique<ModK>(k)); break;
            case 6: ops.push_back(std::make_unique<FlipK>(k)); break;
            default: ops.push_back(std::make_unique<MixK>(k)); break;
        }
    }
    return ops;
}

int main() {
    const int64_t reps = (8000 / SCALE) < 1 ? 1 : (8000 / SCALE);
    std::vector<std::unique_ptr<Op>> ops = build(OPS);
    int64_t acc = 1;
    int64_t check = 0;
    for (int64_t r = 0; r < reps; r++) {
        acc = (acc + r) % M;
        for (const std::unique_ptr<Op> &op : ops) acc = op->apply(acc);
        check = (check + acc) % M;
    }
    std::printf("%d\n", (int)ops.size());
    std::printf("%lld\n", (long long)acc);
    std::printf("%lld\n", (long long)check);
    return 0;
}
